### PR TITLE
[core] Internal kv support in gcs

### LINKED ISSUE #14657: Move Internal KV to GCS

### PR DESCRIPTION
**Pull request by fishbone**
## Why are these changes needed?
To get rid of the dependence of Redis, internal kv needs to be migrated to GCS to be hidden from end-users. 
In this PR, a `KVManager` is introduced in GCS and the call chain from GCS to core_worker is added.

## Related issue number

Closes #14657
## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by ericl — COMMENTED** [review]
@iycheng could you also add the changes to internal_kv.py to use this new GCS client and see if it passes tests?

One issue I can see is that monitor.py uses internal_kv, however monitor.py doesn't have a core worker instance. So it might be better to directly RPC to the GCS client from with a gRPC client in python land.

File: python/ray/_raylet.pyx
Comment (rkooo567) [inline python/ray/_raylet.pyx:946]: This means we can use these only within ray workers (driver + worker)? As far as I remember, there are components that are not parts of Ray core components but use these values.

**Review by rkooo567 — COMMENTED** [review]
I think we need
1. Tests (Python test)
2. Actual refactoring? Are you planning to do them in a separate PR?

**Comment by iycheng** [comment]
Sure, I'll add API for directly calling from GCS. I forgot to add the newly added files which including the test case. Another question is that, do we still need the API at core-worker level? @ericl @rkooo567

**Review by clarkzinzow — COMMENTED** [review]
A lot of this looks great! I would second Eric that using a Python GCS/gRPC client directly would probably be better than relying on the core worker to proxy those RPCs, where those RPCs are hidden underneath the [`internal_kv.py` API](https://github.com/ray-project/ray/blob/master/python/ray/experimental/internal_kv.py). Then the API on the core worker won't be needed.

**Comment by wuisawesome** [comment]
Cc @AmeerHajAli i think this will break backwards compatibility.

**Comment by ericl** [comment]
Could we keep it feature flagged for compatibility reasons?

**Comment by iycheng** [comment]
> Cc @AmeerHajAli i think this will break backwards compatibility.

Could you give me more detail about this?

**Comment by AmeerHajAli** [comment]
> Could we keep it feature flagged for compatibility reasons?

How feature flagging though would solve the issue?  This means we have to run every ray version in the product with the feature flag or force users to use >= ray1.3 which we need to discuss.

**Comment by ericl** [comment]
Or perhaps we can support both paths for a while? Would be good to
enumerate the options here.

File: src/ray/protobuf/gcs_service.proto
Comment (wuisawesome) [inline src/ray/protobuf/gcs_service.proto:517]: I think we need to move this into the autoscaler backwards compatibility section unless we find a way to not break the autoscaler.

File: python/ray/_raylet.pyx
Comment (raulchen) [inline python/ray/_raylet.pyx:938]: I think we need a flag that indicates whether the data is for the current job (e.g., serialized functions). For these data, we need to clean them up when the job finishes.

Comment (iycheng) [reply python/ray/_raylet.pyx:938]: The current implementation which we want to get rid of is used to manage global resources, so that's why it doesn't have scope data.

We can add them based on this PR later.

**Comment by iycheng** [comment]
Somehow the PR is too big :(  Given that we want to merge it asap, maybe it's ok. I can split this one if required.

**Comment by wuisawesome** [comment]
Yeah :(

Maybe we can start with the part that proxies internal kv through GCS (but is still backed by redis)? I think that part is uncontroversial, establishes all the APIs, and still backwards compatible with the autoscaler right?

**Comment by iycheng** [comment]
I'll do another update today, which makes the backend storage consistent to avoid incompatibility

**Comment by clarkzinzow** [comment]
@iycheng FWIW, I'm ok with reviewing this PR at its current size.

**Comment by wuisawesome** [comment]
+1, I think we can live with the PR at its current size if we can ensure backwards compatibility (which it sounds like we can via the prefix thing we talked about offline).

**Comment by iycheng** [comment]
Sorry for the delay. Somehow addicted to the memory profiling issues and make no good progress T_T

I'll update it today

**Comment by iycheng** [comment]
@wuisawesome redis publish  is also used. I guess I also need to expose that API in gcs client.

https://github.com/ray-project/ray/blob/fab2c40d2938b6bbdf47e94475d614403e754dce/python/ray/_private/utils.py#L102

**Comment by wuisawesome** [comment]
it's not relevant to kv store though right? we definitely will have to handle the logging code soon though.

**Comment by iycheng** [comment]
> it's not relevant to kv store though right? we definitely will have to handle the logging code soon though.

Sure. I thought that we want to remove redis in python. I'll leave the unrelated parts there then.

**Comment by wuisawesome** [comment]
we do, but the logging stuff is redis pubsub, and we probably don't want to implement pubsub in this pr

**Comment by iycheng** [comment]
> we do, but the logging stuff is redis pubsub, and we probably don't want to implement pubsub in this pr

Agreed.

**Comment by ericl** [comment]
This is a release blocker since it impacts forward compatibility of autoscaler in separate container.

File: python/ray/includes/common.pxd
Comment (wuisawesome) [inline python/ray/includes/common.pxd:170]: I think these changes are unnecessary?

**Review by wuisawesome — CHANGES_REQUESTED** [review]
Looks like we're pretty close. 

1. For `monitor.py` changes, I think we need some solution that doesn't break with the combination: 1.2.0 cluster + 1.3.0 autoscaler for now. 
2. Looks like there are some UTIL_WORKER things polluting the diff (presumably this is related to runtime environments and not this PR?) 
3. Minor misc comments

File: python/ray/includes/common.pxd
Comment (iycheng) [reply python/ray/includes/common.pxd:170]: Let me merge to upstream/master. I have another fix for this one which just merged.

File: python/ray/__init__.py
Comment (iycheng) [reply python/ray/__init__.py:107]: you are right. let me remove this.

File: python/ray/_private/monitor.py
Comment (iycheng) [reply python/ray/_private/monitor.py:100]: I think we'll have gcs running here, right? In L105, it's gcs_channel. If we have gcs running, it should work.

File: python/ray/includes/gcs_client.pxd
Comment (iycheng) [reply python/ray/includes/gcs_client.pxd:32]: I thought about putting it in a cpp file, but then realized that it's only for python only.
So I think that maybe I should just put it here.

File: python/ray/_raylet.pyx
Comment (clarkzinzow) [inline python/ray/_raylet.pyx:855]: Nit:
```suggestion
def disconnect_from_gcs(gcs_client):
```

File: python/ray/_raylet.pyx
Comment (clarkzinzow) [inline python/ray/_raylet.pyx:851]: We should use the logger instead of plain print statements.

```suggestion
    logger.info(f"Connecting to gcs: {ip}, {port}, {password}")
```

File: python/ray/_private/monitor.py
Comment (clarkzinzow) [inline python/ray/_private/monitor.py:125]: What happens if the GCS client is garbage collected without disconnecting? Couldn't that disconnecting logic go into the `__dealloc__` method in the Cython `GcsClient` class or something like that?

File: python/ray/includes/gcs_client.pxd
Comment (clarkzinzow) [reply python/ray/includes/gcs_client.pxd:32]: +1 on a separate cpp file, but I don't think that should be a blocker for this being merged.

File: python/ray/tests/test_multi_node_2.py
Comment (clarkzinzow) [inline python/ray/tests/test_multi_node_2.py:225]: Why is this necessary? Seems odd.

File: src/ray/gcs/accessor.h
Comment (clarkzinzow) [inline src/ray/gcs/accessor.h:796]: Nit: I don't know if it's best practice, but I'd usually vote for a pointer for this kind of output parameter, e.g.
```suggestion
  Status Get(const std::string &key, std::string *value_out);
```

File: src/ray/gcs/accessor.h
Comment (clarkzinzow) [inline src/ray/gcs/accessor.h:779]: These functions should probably have documenting comments like the other accessor APIs, even though these functions are very simple.

File: src/ray/protobuf/gcs_service.proto
Comment (clarkzinzow) [inline src/ray/protobuf/gcs_service.proto:540]: Why is this a `oneof`?

**Review by clarkzinzow — COMMENTED** [review]
A few nits and questions

File: python/ray/includes/gcs_client.pxd
Comment (iycheng) [reply python/ray/includes/gcs_client.pxd:32]: Here is that this is more or less cython only code, so that's why I'm putting it here.
I'm not against it, but any comment on why do you prefer a separate cpp file?

File: python/ray/_raylet.pyx
Comment (iycheng) [reply python/ray/_raylet.pyx:851]: Oh, I actually need to remove it. It's debugging code. It'll print in cpp side for connecting I think.

File: python/ray/_private/monitor.py
Comment (iycheng) [reply python/ray/_private/monitor.py:125]: That's OK, the desctructor will be called directly. I believe I already have destructor there.

This del is actually a little bit tricky. Here, the owner of the client is monitor, but it's set in another place (core_worker). When we stop the system, we don't want another thread is referencing gcs which will give seg fault. 

This is for a safe stop.

File: python/ray/tests/test_multi_node_2.py
Comment (iycheng) [reply python/ray/tests/test_multi_node_2.py:225]: It's referencing gcs, and we'd like it to stop monitor before stop gcs.

File: src/ray/gcs/accessor.h
Comment (iycheng) [reply src/ray/gcs/accessor.h:796]: I'll vote 'passing by reference'. In this way, you know the value is not null for sure. If it's input parameter, it'll be const reference, otherwise, just reference. In case you need the 'optional' kind of semantic, I'll pass it by the pointer.

Do you have some comments that why you prefer a pointer?

File: src/ray/protobuf/gcs_service.proto
Comment (iycheng) [reply src/ray/protobuf/gcs_service.proto:540]: In case you are getting non-existing value, it should be none. status here is for other usage like communication error I think?

File: src/ray/gcs/accessor.h
Comment (iycheng) [reply src/ray/gcs/accessor.h:779]: Sure, I'll follow the convention.

Well, sometimes, I feel that we're adding too many not useful comments, like the comments which are not necessary, for example [here](https://github.com/ray-project/ray/blob/3869c9d342a425bc48f9c8df30ce1be02e633cda/src/ray/gcs/gcs_server/gcs_server.cc#L74).

File: src/ray/protobuf/gcs_service.proto
Comment (clarkzinzow) [reply src/ray/protobuf/gcs_service.proto:540]: `GcsStatus` is actually for application-level `ray::Status` statuses, e.g. https://github.com/ray-project/ray/blob/9ceb0ca6410bb72e6c273ae8fe7f7af57d6649ba/src/ray/gcs/gcs_client/service_based_accessor.cc#L230-L234, and I think using that with a `ray::Status::NotFound()` status would be better than a pseudo-oneof in my opinion.

File: src/ray/gcs/accessor.h
Comment (clarkzinzow) [reply src/ray/gcs/accessor.h:796]: I've generally preferred a pointer since it makes it more obvious at the _call site_ that it's an output parameter:
```
std::string out;
Get("foo", out);
// vs.
Get("foo", &out);
```
However, after a quick read of a few different style guides and Q&As, I agree that since optional semantics aren't needed, passing by reference is the way to go.

File: src/ray/protobuf/gcs_service.proto
Comment (iycheng) [reply src/ray/protobuf/gcs_service.proto:540]: I see. Thanks for the correction. Let me check how to update it. I agree status is a better choice. I was thinking that it's for other purposes.

File: src/ray/gcs/accessor.h
Comment (clarkzinzow) [reply src/ray/gcs/accessor.h:779]: I completely agree about those kind of comments, although I'm a bit of a stickler about documenting abstraction interfaces, even if the semantics are pretty obvious from the function name and signature. 😛

File: python/ray/experimental/internal_kv.py
Comment (clarkzinzow) [inline python/ray/experimental/internal_kv.py:8]: Also, this `target` argument doesn't appear to be used.

Comment (iycheng) [reply python/ray/experimental/internal_kv.py:8]: Good catch! I forgot to remove this when fixing lifecycle issues!

File: src/ray/gcs/accessor.h
Comment (iycheng) [reply src/ray/gcs/accessor.h:779]: Yes, I'm good with API-level documentation. It's necessary.

**Comment by clarkzinzow** [comment]
A more meta comment (probably for a future PR) on tying the internal KV implementation to the core worker:

With supporting a Python-land GCS client independent of the core worker being a first-class use case (e.g. in the autoscaler), it seems as if patching on an externally construct GCS client to a dummy core worker is not only a bit gross in terms of code, but also adds complexity in ensuring that the client properly disconnects upon destruction. I wonder if it would be better to tweak the internal KV implementation a bit to use a global GCS client singleton that can come from either the core worker or a manually constructed GCS client:

```python
_gcs_client = None

def _get_gcs_client():
    global _gcs_client
    if _gcs_client is None:
        worker = ray.worker.global_worker
        if getattr(worker, "mode", None) is not None:
            _gcs_client = worker.gcs_client
    return _gcs_cliient

def _initialize_gcs_client(ip, port, password):
    global _gcs_client
    _gcs_client = GcsClient.make_from_address(ip, port, password)

@client_mode_hook
def _internal_kv_initialized(target=None):
    return _gcs_client is not None

@client_mode_hook
def _internal_kv_get(key: Union[str, bytes]) -> bytes:
    """Fetch the value of a binary key."""
    return _get_gcs_client().kv_get(key)
```

We would no longer need this [dummy worker in the autoscaler](https://github.com/ray-project/ray/blob/9ceb0ca6410bb72e6c273ae8fe7f7af57d6649ba/python/ray/_private/monitor.py#L109-L113), which would be replaced by an `_initialize_gcs_client(ip, port, password)` call. I _think_ this should also obviate the need for special destruction handling in the autoscaler or elsewhere.

This can wait for a future PR, but I do think that it would clean up the Python internal KV implementation and usage.

File: python/ray/includes/gcs_client.pxd
Comment (wuisawesome) [reply python/ray/includes/gcs_client.pxd:32]: I think it's more of a preference to keep the cython layer thin because we don't have a lot of cython expertise, and we tend to treat it as a "glue" layer, so there's very little test coverage. 

> I don't think that should be a blocker for this being merged.

+1

**Comment by iycheng** [comment]
@clarkzinzow I agreed with you. I actually did it this way to get rid of the hacky worker stuff. The `target` you saw in the code was for this. But there is one issue that I was not fixed at that time about the lifecycle.  :(

We can set the client from either worker.py or monitor.py, and they have a different lifecycle. The handler is a global kind of thing, so which one should overwrite which is a problem here.

I'm thinking about whether we should have an internal kv class and monitor will use that specifically. For the rest of the places, we might just use the normal one. Since I realized I might need more time on this, I just reverted it back to the original way of doing this.

File: python/ray/includes/gcs_client.pxd
Comment (iycheng) [reply python/ray/includes/gcs_client.pxd:32]: Got it 👍

File: python/ray/_private/monitor.py
Comment (wuisawesome) [reply python/ray/_private/monitor.py:100]: But it will be the 1.2.0 GCS server, so how will it know about the `InternalKVGcsService`?

**Comment by wuisawesome** [comment]
Bump on the monitor compatibility question. 

@clarkzinzow can you approve when you're happy with it?

**Comment by iycheng** [comment]
I got the crash because of the destructing ordering problem. 
When the test case finished, it'll stop gcs. And due to gcs_client, it'll crash by complaining failed to reconnect to gcs.
To make it simple, we need to have a way, not RAY_CHECK, but doing something else.

Monitor use gcs_client by running it on a separate thread, which is the problem.

File: python/ray/_private/monitor.py
Comment (clarkzinzow) [reply python/ray/_private/monitor.py:100]: What's the plan here for autoscaler compatibility with old clusters? The only solution that comes to mind is to preserve the existing Redis-based internal KV and fall back to that in the autoscaler if a test ping through the internal KV accessor fails.

File: python/ray/includes/gcs_client.pxd
Comment (clarkzinzow) [inline python/ray/includes/gcs_client.pxd:50]:
```suggestion
          RAY_LOG(DEBUG) << "Destructing GCS client and associated event loop thread.";
```

File: python/ray/includes/gcs_client.pxd
Comment (clarkzinzow) [inline python/ray/includes/gcs_client.pxd:47]: Nit: A bit more idiomatic.
```suggestion
           RAY_CHECK_OK(Connect(io_context_));
```

**Review by clarkzinzow — APPROVED** [review]
LGTM overall except for the autoscaler compatibility issue and a log message in the GCS client destructor. Also:

> I got the crash because of the destructing ordering problem.
> When the test case finished, it'll stop gcs. And due to gcs_client, it'll crash by complaining failed to reconnect to gcs.
> To make it simple, we need to have a way, not RAY_CHECK, but doing something else.
>
> Monitor use gcs_client by running it on a separate thread, which is the problem.

My guess is that the autoscaler needs shutdown logic to explicitly disconnect and then destroy the GCS client, similar to what the [core worker](https://github.com/ray-project/ray/blob/599f325ab550950c8e765b6043e503678e5bcf70/src/ray/core_worker/core_worker.cc#L811) and [raylet](https://github.com/ray-project/ray/blob/599f325ab550950c8e765b6043e503678e5bcf70/src/ray/raylet/main.cc#L256) does with its GCS client. And I'm wondering if it would be better if the IO context had to be provided to the GCS client, where the event loop thread and IO context  lifecycles are managed by the caller... I haven't thought about that enough to know that would work better.

File: python/ray/_private/monitor.py
Comment (wuisawesome) [reply python/ray/_private/monitor.py:100]: I think we can just revert the changes to this file entirely, then once we cut off support for 1.2.0, do this + flip the flag to use in-memory store together.

File: python/ray/includes/gcs_client.pxd
Comment (iycheng) [reply python/ray/includes/gcs_client.pxd:50]: Ah, sorry I left the debug code here again. Thanks for the catching!

**Comment by iycheng** [comment]
> LGTM overall except for the autoscaler compatibility issue and a log message in the GCS client destructor. Also:
> 
> > I got the crash because of the destructing ordering problem.
> > When the test case finished, it'll stop gcs. And due to gcs_client, it'll crash by complaining failed to reconnect to gcs.
> > To make it simple, we need to have a way, not RAY_CHECK, but doing something else.
> > Monitor use gcs_client by running it on a separate thread, which is the problem.
> 
> My guess is that the autoscaler needs shutdown logic to explicitly disconnect and then destroy the GCS client, similar to what the [core worker](https://github.com/ray-project/ray/blob/599f325ab550950c8e765b6043e503678e5bcf70/src/ray/core_worker/core_worker.cc#L811) and [raylet](https://github.com/ray-project/ray/blob/599f325ab550950c8e765b6043e503678e5bcf70/src/ray/raylet/main.cc#L256) does with its GCS client. And I'm wondering if it would be better if the IO context had to be provided to the GCS client, where the event loop thread and IO context lifecycles are managed by the caller... I haven't thought about that enough to know that would work better.

Actually, for monitor, the io_pool is owned by itself (gcs_client in cyhton), since there is no core_worker at that time.
I think it's more like, the current code destructing everything first (also gcs), but not the monitor, and monitor's gcs_client still have thread running (from io_context) there, want to reconnect. Eventually, it'll fail and crash (due to RAY_CHECK).

In this case, we need to ensure gcs is the last one shutdown.

File: python/ray/_private/monitor.py
Comment (iycheng) [reply python/ray/_private/monitor.py:100]: > But it will be the 1.2.0 GCS server, so how will it know about the InternalKVGcsService?
Do you mean autoscaler and ray will share the same gcs? Then I understood it incorrectly. 

This will only work if autoscaler uses (autoscaler doesn't upgrade) the ray GCS.

**Comment by clarkzinzow** [comment]
> Actually, for monitor, the io_pool is owned by itself (gcs_client in cyhton), since there is no core_worker at that time.
> I think it's more like, the current code destructing everything first (also gcs), but not the monitor, and monitor's gcs_client still have thread running (from io_context) there, want to reconnect. Eventually, it'll fail and crash (due to RAY_CHECK).
> 
> In this case, we need to ensure gcs is the last one shutdown.

Isn't the solution there to ensure that the autoscaler (monitor) is disconnecting, shutting down and being properly destructed? I feel like this is more an issue with the autoscaler and/or the GCS client not properly disconnecting and shutting down than the timing of the GCS shutting down, which I don't think we'll be able to control in general. I don't think that the GCS waits for all raylets to shut down before shutting itself down, yet I don't think that the raylets have this problem with their GCS client. I might be missing something though.

**Comment by iycheng** [comment]
> > Actually, for monitor, the io_pool is owned by itself (gcs_client in cyhton), since there is no core_worker at that time.
> > I think it's more like, the current code destructing everything first (also gcs), but not the monitor, and monitor's gcs_client still have thread running (from io_context) there, want to reconnect. Eventually, it'll fail and crash (due to RAY_CHECK).
> > In this case, we need to ensure gcs is the last one shutdown.
> 
> Isn't the solution there to ensure that the GCS client used by the autoscaler (monitor) is disconnecting, shutting down and being properly destructed? I feel like this is more an issue with the autoscaler and/or the GCS client not properly disconnecting and shutting down than the timing of the GCS shutting down, which I don't think we'll be able to control in general. I don't think that the GCS waits for all raylets to shut down before shutting itself down, yet I don't think that the raylets have this problem with their GCS client. I might be missing something though.

You understand it correctly. But the issue here is more related to the test case I think which we still can control. If it's normal use case, and the GCS crashed. It's OK for components to crash to prevent errors from happening. 

While in tests, we don't want crash, otherwise, it'll be marked as a failure which is not. We either make sure gcs is the last one stopped in test or we update [this](https://github.com/ray-project/ray/blob/493d15e05beb0e1bf41a9f18cb68d6d38d89ce48/src/ray/gcs/gcs_client/service_based_gcs_client.cc#L228-L230)

**Comment by amogkam** [comment]
@iycheng is there an ETA on this PR since it is a release blocker? Thanks.

**Comment by amogkam** [comment]
Also, are there any release tests that need to be re-run after this PR is merged? @iycheng @ericl

**Comment by wuisawesome** [comment]
@amogkam can you rerun the scalability envelope (to test the autoscaler) and any multi-node test using the releaser (to test one-way compatibility)?

**Comment by amogkam** [comment]
@wuisawesome OK will do after this gets merged.

**Comment by iycheng** [comment]
I just updated this one to fix the compatible issue. Let me resolve the conflict and update it. I'm wondering whether we need manually testing on autoscaler to make sure it's working @wuisawesome

**Comment by wuisawesome** [comment]
Once we get the merge conflict fixed/get the wheel, let's sync up offline about how to do some testing for this. I have some ideas, but we don't have a formal structure for this type of compatibility testing yet.

File: python/ray/_private/monitor.py
Comment (AmeerHajAli) [inline python/ray/_private/monitor.py:31]: since we will be removing the ray.* imports from these files, is there a way this can be in ray.autoscaler.*?

Comment (AmeerHajAli) [reply python/ray/_private/monitor.py:31]: Or can we actually make this grpc based?

Comment (iycheng) [reply python/ray/_private/monitor.py]: Actually connect_to_gcs and disconnect_from_gcs are built on gcs directly.

We don't use grpc based method instead of having a wrapper on cpp gcs is because it can share the same connection if we actually have a worker there. Since we already have the issue that there are too many connections, we probably don't want it here. 

Since internal_kv is not decoupled from ray, it makes sense to import it from ray, right?

**Comment by iycheng** [comment]
The failed test passed locally after the merge to upstream. I'll give it another try.

**Comment by iycheng** [comment]
@wuisawesome test looks good now.

**Review by wuisawesome — APPROVED** [review]

**Pull request merged**
